File: crates/thicket-net/src/identity.rs
//! Local identity bundle and verified-peer handle for a channel.

use thicket_core::{Id, KeyEndorsement, RootKey, WorkingKey};

/// Everything a node presents to authenticate itself on a channel: its id, the
/// root public key (so the peer can check `id == hash(root)`), the working-key
/// endorsements, and the working key it signs with.
///
/// `Clone` so a single identity can drive many concurrent connections (e.g. a
/// server accepting channels).
#[derive(Clone)]
pub struct LocalIdentity {
    pub id: Id,
    pub root_public_key: Vec<u8>,
    pub endorsements: Vec<KeyEndorsement>,
    pub working: WorkingKey,
}

impl LocalIdentity {
    /// Build a local identity from a root key, generating and endorsing a fresh
    /// working key valid for `valid_secs` from now.
    pub fn from_root(root: &RootKey, valid_secs: u64) -> Self {
        let working = WorkingKey::generate();
        let now = crate::unix_now();
        let endorsement = root
            .endorse(&working.public(), 0, now + valid_secs)
            .expect("endorse working key");
        LocalIdentity {
            id: root.id(),
            root_public_key: root.public(),
            endorsements: vec![endorsement],
            working,
        }
    }
}

/// A peer whose identity was authenticated during the handshake.
#[derive(Clone, Debug)]
pub struct VerifiedPeer {
    pub id: Id,
    pub working_pub: Vec<u8>,
    /// Expiry of the peer's endorsed working key, captured at handshake. Used to
    /// reject messages once the key is no longer valid (per-message freshness).
    pub key_not_after: u64,
}

File: sdk/py/thicket/identity.py
"""Local identity bundle for connecting/serving."""

from __future__ import annotations

import time

from . import crypto


def unix_now() -> int:
    return int(time.time())


class LocalIdentity:
    def __init__(self, root: crypto.RootKey, working: crypto.WorkingKey, endorsements):
        self.root = root
        self.working = working
        self.endorsements = endorsements
        self.id = root.id()
        self.root_public_key = root.public()

    @classmethod
    def from_root(cls, root: crypto.RootKey, valid_secs: int = 1_000_000_000) -> "LocalIdentity":
        working = crypto.WorkingKey.generate()
        endorsement = root.endorse(working.public(), 0, unix_now() + valid_secs)
        return cls(root, working, [endorsement])

    @classmethod
    def generate(cls) -> "LocalIdentity":
        return cls.from_root(crypto.RootKey.generate())
